Next/previous background keyboard shortcuts

vanreece opened the pull request:
Binds , to previous background in list, and . to next background in
list. The thinking to my mind is that those are under the < and >
characters on a US keyboard, but that may not be international enough.

Speaking of which, I put those keys in core.yaml (en > background >
(next|previous)_background > key). Not really sure if that's the best
place for it or not, open to suggestions. Also, do we commit the built
json file "dist/locales/en.json" or not?

This takes the current list of backgrounds, in displayed order and
allows you to iterate through it forwards or backwards. It will only do
the custom background if there's a custom background defined.

Bonkles commented on modules/ui/background.js:
```diff
@@ -318,6 +316,45 @@ export function uiBackground(context) {
+            .filter(filter);
+    }
+
+    function chooseBackgroundAtOffset(offset) {
+        var backgrounds = getBackgrounds(function(d) { return !d.isHidden() && !d.overlay; });
+        backgrounds.sort(sortSources);
+        var currentBackground = context.background().baseLayerSource();
+        var foundIndex = -1;
+        for (let index = 0; index < backgrounds.length; index++) {
+            if (backgrounds[index] === currentBackground) {
+                foundIndex = index;
+                break;
+            }
+        }
+        if (foundIndex === -1) {
+            // Can't find the current background, so just do nothing
+            return;
+        }
+
+        var nextBackgroundIndex = (foundIndex + offset + backgrounds.length) % backgrounds.length;
+        var nextBackground = backgrounds[nextBackgroundIndex];
+        if (nextBackground.id === 'custom' && !nextBackground.template()) {
+            nextBackgroundIndex = (nextBackgroundIndex + offset + backgrounds.length) % backgrounds.length;
+            nextBackground = backgrounds[nextBackgroundIndex];
+        }
+        chooseBackground(nextBackground);
+    }
```
This function uses `var` and `let`. There are plenty of places where 'var' is still in the iD code, but any new code should be as ES6-ified as possible. Replace `var` with `const` or `let` as appropriate!

vanreece replied:
![image](https://user-images.githubusercontent.com/1225959/86977859-62e63100-c132-11ea-9183-071b3ce783c6.png)

Bonkles replied:
yasssssss

Bonkles commented on modules/ui/background.js:
```diff
@@ -318,6 +316,45 @@ export function uiBackground(context) {
+        var backgrounds = getBackgrounds(function(d) { return !d.isHidden() && !d.overlay; });
+        backgrounds.sort(sortSources);
+        var currentBackground = context.background().baseLayerSource();
+        var foundIndex = -1;
+        for (let index = 0; index < backgrounds.length; index++) {
+            if (backgrounds[index] === currentBackground) {
+                foundIndex = index;
+                break;
+            }
+        }
```
This entire block can be replaced with a call of the array helper method `indexOf()` - it returns -1 if not found: 

`        const foundIndex = backgrounds.indexOf(currentBackground);`

vanreece replied:
Sweet javascripty goodness. Thanks for that. :)

bhousel commented on data/core.yaml:
```diff
@@ -735,6 +735,10 @@ en:
       tooltip: Show advanced background information.
     fix_misalignment: Adjust imagery offset
     offset: "Drag anywhere in the gray area below to adjust the imagery offset, or enter the offset values in meters."
+    next_background:
```
whoa cool - I never noticed the RapiD badges here..

vanreece replied:
Oh yeah! shortcuts.json. I knew I'd forget something.

vanreece commented:
@Bonkles Okay, incorporated feedback. Looking pretty good to me. I added the RapiD badge to the shortcuts, I guess I could try to land a similar change in iD and then eventually we could remove that badge if all goes well there, eh?

Bonkles commented:
> @Bonkles Okay, incorporated feedback. Looking pretty good to me. I added the RapiD badge to the shortcuts, I guess I could try to land a similar change in iD and then eventually we could remove that badge if all goes well there, eh?

Yup, that sounds like a good way to play it! Thanks for addressing feedback too!

Bonkles commented:
@vanreece what are your next steps with this PR? Get it merged in, test with internal testers in an internal build, then iterate? or is there more work you'd like to do on it?

Bonkles commented on modules/ui/background.js:
```diff
@@ -318,6 +316,39 @@ export function uiBackground(context) {
+        const backgrounds = getBackgrounds(function(d) { return !d.isHidden() && !d.overlay; });
+        backgrounds.sort(sortSources);
+        const currentBackground = context.background().baseLayerSource();
+        const foundIndex = backgrounds.indexOf(currentBackground);
```
So terse. So *clean*. It makes my heart sing.

vanreece commented:
@Bonkles okay, I cherry picked the two fixes that @bhousel had on the esri branch, which fixes the tests on master and here into this commit. How do I kick the CI

Bonkles commented on css/80_app_fb.css:
```diff
@@ -472,4 +472,31 @@ g.layer-ai-features.hide-rapid-buildings path.building  {
     background-image: url('img/tag-sources.gif'); 
+}
+
+ul.layer-list.layer-background-list li button.background-favorite-button{
```
format nit- add a space between the .css rule specifiers and the opening brace ⬇️ 

`ul.layer-list.layer-background-list li button.background-favorite-button {`

Bonkles commented on data/core.yaml:
```diff
@@ -2110,6 +2114,8 @@ en:
         title: "Display options"
         background: "Toggle background imagery pane"
         background_switch: "Switch back to last background"
+        background_previous: "Switch to previous background in list"
```
Consider removing 'in list' from this description- I think 'switch to previous background' is perfectly adequate

Bonkles commented on modules/ui/background.js:
```diff
@@ -318,6 +354,39 @@ export function uiBackground(context) {
+            .sources(context.map().extent(), context.map().zoom(), true)
+            .filter(filter);
+    }
+
```
Yet another formatting nit- functions in iD have two spaces between them. NOT a big deal, but something to note!
